lib/patch/functions.js

```diff
@@ -48,6 +48,58 @@ function patchSetClearFunction(obj, fnNames) {
   });
 };
 
+
+/**
+ * Patch a function that may be called again from within the callback that it executes.
+ *
+ *  1) Calculate the function signature of the given callback, and keep a reference count
+ *     on it.
+ *  2) If the patched function is called from within its own callback function, bind to the
+ *     parent zone, and then increment the reference count.
+ *  3) After the callback function has executed, decrement the reference count.
+ *  4) If the reference count is <= 0, clear the cached function signature and dequeue
+ *     the current task.
+ */
+function patchRepeatingCallbackFunction(obj, fnNames) {
+  fnNames.forEach(function (name) {
+    var delegate = obj[name];
+    if (delegate) {
+      var ids = {};
+      global.zone[name] = function (fn) {
+        if(fn){
+          // Reference count functions by their toString() value.
+          var fnRef = fn, id = fn.toString();
+          arguments[0] = function () {
+            fnRef.apply(this, arguments);
+            ids[id]--;
+            if (ids[id] <= 0) {
+              delete ids[id];
+              zone.dequeueTask();
```

Comment:
I'm still a bit unsure about how the `dequeueTask` method is supposed to work, because it is unimplemented in the base zone implementation.  Thanks for the correction.